Rewrite the termination and solution statuses

joehuchette opened the pull request:
They are now "MOI light". This was semi-automated, so a translation mistake may have slipped through.

mtanneau commented on src/lp/dominated_column.jl:
```diff
@@ -43,8 +43,8 @@ function remove_dominated_column!(ps::PresolveData{T}, j::Int; tol::T=100 * sqrt
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/lp/dominated_column.jl:
```diff
@@ -100,8 +100,8 @@ function remove_dominated_column!(ps::PresolveData{T}, j::Int; tol::T=100 * sqrt
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/lp/empty_column.jl:
```diff
@@ -37,8 +37,8 @@ function remove_empty_column!(ps::PresolveData{T}, j::Int) where {T}
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/lp/empty_column.jl:
```diff
@@ -69,8 +69,8 @@ function remove_empty_column!(ps::PresolveData{T}, j::Int) where {T}
-            ps.solution.primal_status = Sln_InfeasibilityCertificate
-            ps.solution.dual_status = Sln_Unknown
+            ps.solution.primal_status = INFEASIBILITY_CERTIFICATE
+            ps.solution.dual_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
 
     # TODO: check status for potential early return
-    ps.status == Trm_Unknown || return ps.status
+    ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
 
```
We need to switch `ps.status` to `IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
     ps.updated = true
     npasses = 0  # TODO: maximum number of passes
-    while ps.updated && ps.status == Trm_Unknown
+    while ps.updated && ps.status == NOT_CALLED
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
 
         bounds_consistency_checks!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
+        ps.status == NOT_CALLED || return ps.status
         remove_empty_columns!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
         # TODO: remove empty variables as well
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
+        ps.status == NOT_CALLED || return ps.status
         remove_fixed_variables!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
         # Remove forcing & dominated constraints
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
+        ps.status == NOT_CALLED || return ps.status
         remove_forcing_rows!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
         # Remove free and implied free column singletons
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
+        ps.status == NOT_CALLED || return ps.status
         remove_free_column_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
         # Dual reductions
         remove_row_singletons!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -364,59 +364,59 @@ function presolve!(ps::PresolveData{T}) where {T}
+        ps.status == NOT_CALLED || return ps.status
         remove_dominated_columns!(ps)
-        ps.status == Trm_Unknown || return ps.status
+        ps.status == NOT_CALLED || return ps.status
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -507,8 +507,8 @@ function bounds_consistency_checks!(ps::PresolveData{T}) where {T}
             # Farkas ray: y⁺_i = y⁻_i = 1 (any > 0 value works)
-            ps.solution.primal_status = Sln_Unknown
-            ps.solution.dual_status = Sln_InfeasibilityCertificate
+            ps.solution.primal_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/presolve.jl:
```diff
@@ -536,8 +536,8 @@ function bounds_consistency_checks!(ps::PresolveData{T}) where {T}
             # Farkas ray: y⁺_i = y⁻_i = 1 (any > 0 value works)
-            ps.solution.primal_status = Sln_Unknown
-            ps.solution.dual_status = Sln_InfeasibilityCertificate
+            ps.solution.primal_status = NO_POINT
```
`NO_SOLUTION`

mtanneau commented on src/presolve.jl:
```diff
@@ -584,7 +584,7 @@ If an empty column is created later, it is removed on the spot.
     for j in 1:ps.pb0.nvar
         remove_empty_column!(ps, j)
-        ps.status == Trm_Unknown || break
+        ps.status == NOT_CALLED || break
```
`IN_PROGRESS`

mtanneau commented on src/presolve.jl:
```diff
@@ -693,7 +693,7 @@ function remove_dominated_columns!(ps::PresolveData{T}) where {T}
     for (j, flag) in enumerate(ps.colflag)
         remove_dominated_column!(ps, j)
-        ps.status == Trm_Unknown || break
+        ps.status == NOT_CALLED || break
```
`IN_PROGRESS`

mtanneau commented on test/lp/empty_column.jl:
```diff
@@ -38,25 +38,25 @@ function emtpy_column_tests(T::Type)
                 @test sol.x[1] > 0
             else
-                @test ps.status == MathOptPresolve.Trm_Unknown
+                @test ps.status == MOP.NOT_CALLED
```
`REDUCED`

mtanneau commented on test/lp/empty_row.jl:
```diff
@@ -27,7 +27,7 @@ function empty_row_tests(T::Type)
 
     @test ps.updated
-    @test ps.status == MathOptPresolve.Trm_Unknown
+    @test ps.status == MOP.NOT_CALLED
```
`REDUCED`

mtanneau commented on src/solution.jl:
```diff
@@ -18,7 +18,9 @@ mutable struct Solution{T}
     Solution{T}(m, n) where {T} = new{T}(
-        m, n, Sln_Unknown, Sln_Unknown, false, false,
+        m, n,
+        NO_POINT, NO_POINT,
```
`NO_SOLUTION`

joehuchette commented:
OK, I have slightly restructed the ``TerminationStatus``es.

1. I renamed the enum to ``ModelStatus``, since it was also being used internally to track the status of the model as presolve progresses.
2. I lumped together ``NOT_CALLED``, ``IN_PROGRESS``, and ``REDUCED`` into ``NOT_INFERRED``. This means that presolve has not been able to infer the status of the model yet (e.g. optimality, primal infeasibility, etc.). IMO this is really what we're trying to track here, and it is true when presolve has not yet been called, when we are "in progress", and as a return type that means you still need to optimize over the inferred model.

I also added a little helper macro to get rid of all the status checks inside the presolve loop. I think it makes it a bit easier to read, but YMMV.

Let me know what you think.

joehuchette commented:
Lol so it turns out ``presolve!`` isn't being tested, so my macro had a boneheaded escaping error. Fixed, and added a test for ``presolve!``.
